micropip/package_index.py

```diff
@@ -303,36 +308,101 @@ async def query_package(
+                parser = _select_parser(content_type, name, index_base_url=base_url)
+            except ValueError as e:
+                raise ValueError(f"Error trying to decode url: {url}") from e
+            return parser(metadata)
+    # With "unsafe-first-match" or "unsafe-best-match", we need to check all indexes
+    else:
+        for url in index_urls:
+            logger.debug("Looping through index urls: %r", url)
+            if _contain_placeholder(url):
+                url = url.format(package_name=name)
+                logger.debug("Formatting url with package name : %r", url)
+            else:
+                url = f"{url}/{name}/"
+                logger.debug("Url has no placeholder, appending package name : %r", url)
+            try:
+                metadata, headers = await fetch_string_and_headers(url, _fetch_kwargs)
+            except HttpStatusError as e:
+                if e.status_code == 404:
+                    logger.debug("NotFound (404) for %r, trying next index.", url)
+                    continue
+                logger.debug(
+                    "Error fetching %r (%s), trying next index.", url, e.status_code
+                )
+                continue  # try the next index instead of raising
+
+            content_type = headers.get("content-type", "").lower()
+            try:
+                base_url = urlunparse(urlparse(url)._replace(path=""))
+                parser = _select_parser(content_type, name, index_base_url=base_url)
+                projects_info.append(parser(metadata))
+            except ValueError:
+                # Just log and continue with the next index
+                msg = f"Error trying to decode url: {url}"
+                logger.debug(msg)
                 continue
-            logger.debug(
-                "Error fetching %r (%s), trying next index.", url, e.status_code
-            )
-            raise
 
-        content_type = headers.get("content-type", "").lower()
-        try:
-            base_url = urlunparse(urlparse(url)._replace(path=""))
+        if not projects_info:
+            raise ValueError(
+                f"Can't fetch metadata for '{name}'. "
+                "Please make sure you have entered a correct package name "
+                "and correctly specified index_urls (if you changed them)."
+            )
 
-            parser = _select_parser(content_type, name, index_base_url=base_url)
-        except ValueError as e:
-            raise ValueError(f"Error trying to decode url: {url}") from e
-        return parser(metadata)
-    else:
-        raise ValueError(
-            f"Can't fetch metadata for '{name}'. "
-            "Please make sure you have entered a correct package name "
-            "and correctly specified index_urls (if you changed them)."
-        )
+        # For "unsafe-first-match", return the first project that has matching versions
+        if strategy == "unsafe-first-match":
+            return projects_info[0]
```

Comment:
Is it the right implementation? If the `projects_info[0]` does not have the compatible version, I think we should fallback to other indices.

The compatibility check is not handled in this function (at least in the current implementation). So I think so support this strategy, we'll need to modify the Transaction as well.